feat: Add support for paged calls for entities

soeren-schmaljohann-2denker opened the pull request:
BREAKING CHANGE: The `loadEntities` call no longer has a generic parameter that the items will be cast to.

ABausG reviewed (commented):
Okay, I understand now what you are doing.

---

The reason for building the `EntitiesResponse` class in the first place was to have this representing the "page" response.
So I think the page Information like current page, number of items, total pages, etc. should become properties of the `EntitiesResponse` either as parameters or wrapped in an Object that collects the page informations.
The `loadEntities` call should have the desired page as an optional parameter. Also page size and page itself should then be added as function parameters.

---

I like the addition of the `loadNextEntitiesPage` and `loadEntitiesPage` in principal. However adding to the comment from above I think the api could be simplified there to something in the following direction:
```dart
Future<List<EntitiesResponse<T>> loadNextPage({Uri? uri, List<EntitiesResponse> currentPages = [], headers, filters, ...} 
```

However I think this should not necessarily be something in the core package but rather a possible addition for the heinzelmen, because there are some risks involved with this approach as it could lead to duplicated entries when pages change.


---

So to summarize:
I think for a first approach we should slim this down to only extend the loadEntites call and the EntitiesResponse class to accept/provide Page Information. In general I believe managing the page logic, collecting entities should be done by clients implementing it.
Providing additonal logic could be done once we notice that we see patterns that would make sense and are actually needed and then decide if we see them in the core package or rather in the heinzelmen

soeren-schmaljohann-2denker commented:
I integrated the page meta data into the entities response and removed the paging helper functions for now. I also extracted the paging request data into an object, to make it easier to load different pages. Otherwise the user would have to add the index query manually to the uri.

ABausG commented on packages/apptive_grid_core/lib/src/model/entity/entities_response.dart:
```diff
@@ -24,6 +46,68 @@ class EntitiesResponse<T> {
   EntitiesResponse copyWith({List<T>? items}) {
```
The Page Meta Data should also be passed in

soeren-schmaljohann-2denker replied:
done

ABausG commented on packages/apptive_grid_core/lib/src/model/entity/entities_response.dart:
```diff
@@ -24,6 +46,68 @@ class EntitiesResponse<T> {
+  /// The total number of pages available.
+  final int numberOfPages;
+
+  /// The number of entities in each page.
```
Is this for _each_ page or rather for _this_ page? I'd assume if I have a page with 5 items because it's the last page/first page with only that amount it would be lower?

soeren-schmaljohann-2denker replied:
I checked with Cyrille, its for **_each_** page.

ABausG commented on packages/apptive_grid_core/lib/src/network/apptive_grid_client.dart:
```diff
@@ -408,7 +408,7 @@ class ApptiveGridClient extends ChangeNotifier {
     bool isRetry = false,
     Map<String, String> headers = const {},
-    int? pageSize,
+    PagingRequest? pagingRequest,
```
I think it would be fine to just have separate parameters for `page` and `pageSize` here to limit overhead.
Also out of interest is pageSize working without specifying the page itself? That would be useful if for example I only want to display 3 things. So possibly uncoupling the parameters from each other

soeren-schmaljohann-2denker replied:
I did it like this, so the user can't specify just a pageIndex. But after testing it, it even works with only the index set.

To you other question, yes that works that way.

Done

ABausG commented on packages/apptive_grid_core/lib/src/network/apptive_grid_client.dart:
```diff
@@ -763,3 +711,35 @@ class ApptiveGridClient extends ChangeNotifier {
 }
+
+/// Meta data for a paging request
+class PagingRequest {
```
If decided to keep it should be moved out of the client itself and be exposed by the package but I think it is cleaner to separate it

soeren-schmaljohann-2denker replied:
removed
